gems/aws-sdk-core/lib/aws-sdk-core/plugins/user_agent.rb

```diff
@@ -9,38 +9,109 @@ class UserAgent < Seahorse::Client::Plugin
+          def initialize(context)
+            @context = context
+          end
+
+          def to_s
+            ua = sdk_metadata
+            ua += " #{api_metadata}" if api_metadata
+            ua += " #{legacy_api_metadata}" if legacy_api_metadata
+            ua += " #{os_metadata}"
+            ua += " #{legacy_os_metadata}"
+            ua += " #{language_metadata}"
+            ua += " #{legacy_language_metadata}"
+            ua += " #{env_metadata}" if env_metadata
+            ua += " #{feature_metadata}" if feature_metadata
+            ua += " #{config_metadata}" if config_metadata
+            # framework metadata can be added here
+            if @context.config.user_agent_suffix
+              ua += " #{@context.config.user_agent_suffix}"
+            end
+            ua += " #{app_id}" if app_id
+            ua.strip
+          end
+
+          private
+
+          def sdk_metadata
+            "aws-sdk-ruby3/#{CORE_GEM_VERSION}"
+          end
+
+          def api_metadata
+            # TODO: transform serviceId to something?
+            service_id = @context.metadata['serviceId']
+            gem_version = @context[:gem_version]
+            "api/#{service_id}##{gem_version}"
+          end
+
+          def legacy_api_metadata
+            # gem name and gem version is legacy.
+            "#{@context[:gem_name]}/#{@context[:gem_version]}"
+          end
+
+          def os_metadata
+            local = Gem::Platform.local
+            os =
+              case local.os
+              when /mac|darwin/
+                'macos'
+              when /linux|cygwin/
+                'linux'
+              when /mingw|mswin/
+                'windows'
+              else
+                'other'
+              end
+            metadata = "os/#{os}"
+            metadata += "##{local.version}" if local.version
+            metadata += " md/#{local.cpu}" if local.cpu
+            metadata
+          end
+
+          def legacy_os_metadata
+            RUBY_PLATFORM
+          end
 
-          begin
-            ua += " #{RUBY_ENGINE}/#{RUBY_VERSION}"
+          def language_metadata
+            "lang/#{RUBY_ENGINE}##{RUBY_ENGINE_VERSION} md/#{RUBY_VERSION}"
+          end
+
+          def legacy_language_metadata
+            "#{RUBY_ENGINE}/#{RUBY_VERSION}"
           rescue
-            ua += " RUBY_ENGINE_NA/#{RUBY_VERSION}"
+            "RUBY_ENGINE_NA/#{RUBY_VERSION}"
           end
 
-          ua += " #{RUBY_PLATFORM}"
+          def env_metadata
+            return unless (execution_env = ENV['AWS_EXECUTION_ENV'])
 
-          if context[:gem_name] && context[:gem_version]
-            ua += " #{context[:gem_name]}/#{context[:gem_version]}"
+            "exec-env/#{execution_env}"
           end
 
-          if (execution_env = ENV['AWS_EXECUTION_ENV'])
-            ua += " exec-env/#{execution_env}"
+          # TODO: somehow set this for paginators, waiters, resources, s3 transfer, s3 encryption, etc
```

Comment:
Yeah - this gets a bit trickier. For cases where we are creating the client in the HLL, we can set a config option.  However, when passed in... we shouldn't modify the config.  

We may need to add an option to operation methods to take additional options for this.  IE, per operation config?

Reply:
Yeah it's looking like we might need per operation options. Otherwise some features require a config option or a context, but never both.